lib/modules.nix

```diff
@@ -291,19 +333,41 @@ rec {
         in
           if nrOptions == length decls then
             let opt = fixupOptionType loc (mergeOptionDecls loc decls);
-            in evalOptionValue loc opt defns'
+            in {
+              matchedOptions = evalOptionValue loc opt defns';
+              unmatchedDefns = [];
+            }
           else if nrOptions != 0 then
             let
               firstOption = findFirst (m: isOption m.options) "" decls;
               firstNonOption = findFirst (m: !isOption m.options) "" decls;
             in
               throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
           else
+            # Process option sets
             if all (def: isAttrs def.value) defns' then mergeModules' loc decls defns
             else let firstInvalid = findFirst (def: ! isAttrs def.value) null defns';
-            in throw "The option path `${showOption loc}' is an attribute set of options, but it is defined to not be an attribute set in `${firstInvalid.file}'. Did you define its value at the correct and complete path?"
-      ))
-    // { _definedNames = map (m: { inherit (m) file; names = attrNames m.config; }) configs; };
+            in throw "The option path `${showOption loc}' is an attribute set of options, but it is defined to not be an attribute set in `${firstInvalid.file}'. Did you define its value at the correct and complete path?");
+
+      matchedOptions = mapAttrs (n: v: v.matchedOptions) resultsByName;
+
+      # an attrset 'name' => list of unmatched definitions for 'name'
+      unmatchedDefnsByName =
+        # Propagate all unmatched definitions from nested option sets
+        mapAttrs (n: v: v.unmatchedDefns) resultsByName
+        # Plus the definitions for the current prefix that don't have a matching option
+        // removeAttrs defnsByName' (attrNames matchedOptions);
+    in {
+      inherit matchedOptions;
+
+      # Transforms unmatchedDefnsByName into a list of definitions
+      unmatchedDefns = concatLists (mapAttrsToList (name: defs:
+        map (def: def // {
+          # Set this so we know when the definition first left unmatched territory
```

Comment:
You could already modify the `value` as you go.

```
value = { ${name} = def.value; };
```

This removes the need to `map` right before `structuredConfig._module.unstructuredType.merge`. You'd just maintain a consistent `def` all the time.

It may also remove the need for `def.prefix`, depending on whether we need it for error messages.

Reply:
I had it this way initially, without `prefix`, and generating error messages for invalid options by converting `some.invalid.option = ...` to `loc = [ "some" "invalid" "option" ]`. But that doesn't work well with `some.invalid.option = mkDefault ...`, because `loc` would then be e.g. `[ ... "_type" ]`, because `mkDefault` returns an attrset as well, and there's no way for it to know not to recurse into it. I can imagine other cases where something like this happens. So keeping track of the `prefix` doesn't make this a problem.

Also the prefix would be needed to throw a better error message in the case I mentioned above